core/store/models/run_output.go

```diff
@@ -0,0 +1,119 @@
+package models
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/tidwall/gjson"
+	null "gopkg.in/guregu/null.v3"
+)
+
+// RunOutput represents the result of performing a Task
+type RunOutput struct {
+	Data         JSON
+	Status       RunStatus
+	ErrorMessage null.String
+}
+
+// NewRunOutputError returns a new RunOutput with an error
+func NewRunOutputError(err error) RunOutput {
+	return RunOutput{
+		Status:       RunStatusErrored,
+		ErrorMessage: null.StringFrom(err.Error()),
+	}
+}
+
+// NewRunOutputCompleteWithResult returns a new RunOutput that is complete and
+// contains a result
+func NewRunOutputCompleteWithResult(resultVal interface{}) RunOutput {
+	var data JSON
+	data, _ = data.Add("result", resultVal)
+	return NewRunOutputComplete(data)
+}
+
+// NewRunOutputComplete returns a new RunOutput that is complete and contains
+// raw data
+func NewRunOutputComplete(data JSON) RunOutput {
+	return RunOutput{Status: RunStatusCompleted, Data: data}
+}
+
+// NewRunOutputPendingSleep returns a new RunOutput that indicates the task is
+// sleeping
+func NewRunOutputPendingSleep() RunOutput {
+	return RunOutput{Status: RunStatusPendingSleep}
+}
+
+// NewRunOutputPendingConfirmations returns a new RunOutput that indicates the
+// task is pending confirmations
+func NewRunOutputPendingConfirmations() RunOutput {
+	return RunOutput{Status: RunStatusPendingConfirmations}
+}
+
+// NewRunOutputPendingConfirmationsWithData returns a new RunOutput that
+// indicates the task is pending confirmations but also has some data that
+// needs to be fed in on next invocation
+func NewRunOutputPendingConfirmationsWithData(data JSON) RunOutput {
+	return RunOutput{Status: RunStatusPendingConfirmations, Data: data}
+}
+
+// NewRunOutputPendingConnection returns a new RunOutput that indicates the
+// task got disconnected
+func NewRunOutputPendingConnection() RunOutput {
+	return RunOutput{Status: RunStatusPendingConnection}
+}
+
+// NewRunOutputPendingConfirmationsWithData returns a new RunOutput that
+// indicates the task got disconnected but also has some data that needs to be
+// fed in on next invocation
+func NewRunOutputPendingConnectionWithData(data JSON) RunOutput {
+	return RunOutput{Status: RunStatusPendingConnection, Data: data}
+}
+
+// NewRunOutputInProgress returns a new RunOutput that indicates the
+// task is still in progress
+func NewRunOutputInProgress(data JSON) RunOutput {
+	return RunOutput{Status: RunStatusInProgress, Data: data}
+}
+
+// NewRunOutputPendingBridge returns a new RunOutput that indicates the
+// task is still in progress
+func NewRunOutputPendingBridge() RunOutput {
+	return RunOutput{Status: RunStatusPendingBridge}
+}
+
+// HasError returns true if the status is errored or the error message is set
+func (ro RunOutput) HasError() bool {
+	return ro.Status == RunStatusErrored || ro.ErrorMessage.Valid
+}
```

Comment:
Is it valid for a RunOutput to have an error but without status 'Errored'?

Reply:
Technically yes... but I could potentially make the fields private to make absolutely sure.

Reply:
Maybe that would be better?  In my general opinion, if its technically yes, but intentionally no, exploding with a panic would be better (not ideal) than being forgiving about a state which inconsistent.

Reply:
Oh, yeah, to be clear I'm looking at encapsulating it right now, it has the nice side effect of shrinking the interface even further.